armi/utils/asciimaps.py

```diff
@@ -40,236 +60,428 @@ class AsciiMap(object):
+        self.asciiLines = []
+        self._asciiMaxCol = 0
+        for li, line in enumerate(text):
+            columns = line.split()
+            self.asciiLines.append(columns)
+            if len(columns) > self._asciiMaxCol:
+                self._asciiMaxCol = len(columns)
+        self._asciiMaxLine = li + 1  # pylint: disable=undefined-loop-variable
+        self._updateDimensionsFromAsciiLines()
+        self._asciiLinesToIndices()
+        self._makeOffsets()
+        self._updateSlotSizeFromData()
+
+    def _updateSlotSizeFromData(self):
+        """After reading data, update slot size for writing."""
+        slotSize = max(len(v) for v in self.asciiLabelByIndices.values())
+        self._spacer = " " * slotSize
+        fmt = f"{{placeholder:{slotSize}s}}"
+        self._placeholder = fmt.format(placeholder=PLACEHOLDER)
+
+    def _updateDimensionsFromAsciiLines(self):
+        """
+        When converting ascii to data we need to infer the ijMax before reading
+        the ij indices.
+        
+        See Also
+        --------
+        _updateDimensionsFromData : used to infer this information when loading from i,j data
+        """
+        raise NotImplementedError
 
-        Returns
-        -------
-        lattice : dict
-            A mapping between (i,j) indices and specifiers, which can be
-            one or more non-space chars.
+    def _updateDimensionsFromData(self):
+        """
+        After data is loaded, figure out the bounds.
 
         See Also
         --------
-        armi.reactor.grids : Grid objects that this information can be
-            turned into
+        _updateDimensionsFromAsciiLines : used when reading info from ascii lines
         """
-        self.lattice = {}
-        for row, line in enumerate(reversed(text.strip().splitlines())):
-            line = line.strip()
-            base = self._getRowBase(row)
-            for col, char in enumerate(line.split()):
-                i, j = self._getIndices(base, col)
-                self.lattice[i, j] = char
-
-        self.lattice = {k: v for k, v in self.lattice.items() if v != PLACEHOLDER}
-        return self.lattice
-
-    def writeMap(self, stream):
-        """Write an ascii map from the internal lattice structure to a stream."""
-        colI, rowJ = zip(*sorted(self.lattice.keys(), key=lambda ij: ([ij[1], ij[0]])))
-        line = []
-        for i, j in zip(colI, reversed(rowJ)):
-            if line and i == 0:
-                stream.write(" ".join(line) + "\n")
-                line = []
-            line.append(f"{self.lattice.get((i,j), PLACEHOLDER)}")
-        stream.write(" ".join(line) + "\n")
+        self._ijMax = max(sum(key) for key in self.asciiLabelByIndices)
 
     @staticmethod
-    def _getRowBase(row):
-        raise NotImplementedError()
+    def fromReactor(reactor):
+        """
+        Populate mapping from a reactor in preparation of writing out to ascii.
+        """
```

Comment:
Delete, or raise `NotImplementedError`

Reply:
Actually, thinking about this more, I don't think we should be coupling a utils function to something as big as the Reactor class... makes for cyclical dependencies (even without the import, it's still there). Also, it isn't exactly well-specified for a Reactor, since a reactor can have a Core, SFP, assemblies/blocks with their own lattices, etc.

Should probably just remove this. A client will just need to pass `{(i, j): contents for ... in ...}`